akka-sample-sharding-scala/killrweather/src/main/scala/sample/killrweather/WeatherRoutes.scala

```diff
@@ -2,64 +2,92 @@ package sample.killrweather
 
 import scala.concurrent.Future
 import scala.concurrent.duration._
-
 import akka.actor.typed.{ActorRef, ActorSystem}
 import akka.actor.typed.scaladsl.AskPattern._
-import akka.{actor => classic}
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
 import akka.http.scaladsl.model.StatusCodes
 import akka.http.scaladsl.server.Route
+import akka.http.scaladsl.unmarshalling.Unmarshaller
 import akka.util.Timeout
 
-object WeatherRoutes {
-  sealed trait Status extends CborSerializable
-  final case class DataIngested(wsid: String) extends Status
-  final case class WeatherStationAdded(wsid: String) extends Status
-  final case class WeatherStations(ids: Set[String]) extends Status
-  final case class TimeWindow(start: Long, end: Long, value: Double)
-  final case class QueryResult(wsid: String, dataType: String, funcTag: String, readings: Int, value: Vector[TimeWindow]) extends Status
-}
-
 /** HTTP API for
  * 1. Receiving data from remote weather stations
- *    A. Adding a new station when it comes online / goes offline
+ *    A. Adding a new station when it comes online / goes offline (to this node, the state is local)
  *    B. Device samplings over windowed time slices
  * 2. Receiving and responding to queries
  *
- * @param guardian the entry point to the cluster and sharded data aggregates
+ * @param stations the entry point to the cluster and sharded data aggregates
  */
-private[killrweather] final class WeatherRoutes(guardian: ActorRef[Guardian.Command])(implicit system: ActorSystem[_]) {
+private[killrweather] final class WeatherRoutes(stations: ActorRef[Stations.Command], sharding: ClusterSharding)(implicit system: ActorSystem[_]) {
+
+  // timeout used for asking the actor
+  private implicit val timeout: Timeout =
+    system.settings.config.getDuration("killrweather.routes.ask-timeout").toMillis.millis
+
+  private def addWeatherStation(wsid: String): Future[Stations.WeatherStationAdded] =
+    stations.ask(Stations.AddWeatherStation(wsid, _))
 
-  import akka.actor.typed.scaladsl.adapter._
-  implicit val classicSystem: classic.ActorSystem = system.toClassic
-  implicit val timeout: Timeout = system.settings.config.getDuration("killrweather.routes.ask-timeout").toMillis.millis
+  private def recordData(wsid: String, data: WeatherStation.Data): Future[WeatherStation.DataRecorded] = {
+    val ref = sharding.entityRefFor(WeatherStation.TypeKey, wsid)
+    ref.ask(WeatherStation.Record(data, System.currentTimeMillis, _))
+  }
 
+  private def query(wsid: String, dataType: WeatherStation.DataType, function: WeatherStation.Function): Future[WeatherStation.QueryResult] = {
+    val ref = sharding.entityRefFor(WeatherStation.TypeKey, wsid)
+    ref.ask(WeatherStation.Query(wsid, dataType, function, _))
+  }
```

Comment:
Following the Akka HTTP quickstart where we have encapsulated asks into methods to capture the types, allows us to just use `complete(futureCall(param))` when the future completes with a marshallable type